File: Stripe/STPFakeAddPaymentPassViewController.m
//
//  STPFakeAddPaymentPassViewController.m
//  Stripe
//
//  Created by Jack Flintermann on 9/28/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import "STPFakeAddPaymentPassViewController.h"
#import "PKAddPaymentPassRequest+Stripe_Error.h"
#import "StripeError.h"
#import "STPLocalizationUtils.h"

typedef NS_ENUM(NSUInteger, STPFakeAddPaymentPassViewControllerState) {
    STPFakeAddPaymentPassViewControllerStateInitial,
    STPFakeAddPaymentPassViewControllerStateLoading,
    STPFakeAddPaymentPassViewControllerStateError,
    STPFakeAddPaymentPassViewControllerStateSuccess,
};

@interface STPFakeAddPaymentPassViewController ()
@property(nonatomic)PKAddPaymentPassRequestConfiguration *configuration;
@property(nonatomic)STPFakeAddPaymentPassViewControllerState state;
@property(nonatomic)UILabel *contentLabel;
@property(nonatomic)NSString *errorText;
@end

@implementation STPFakeAddPaymentPassViewController

+ (BOOL)canAddPaymentPass {
    return YES;
}

- (nullable instancetype)initWithRequestConfiguration:(PKAddPaymentPassRequestConfiguration *)configuration
                                             delegate:(nullable id<PKAddPaymentPassViewControllerDelegate>)delegate {
    self = [super initWithNibName:nil bundle:nil];
    NSCParameterAssert(delegate);
    if (self) {
        _state = STPFakeAddPaymentPassViewControllerStateInitial;
        _delegate = delegate;
        _configuration = configuration;
        if (!configuration.primaryAccountSuffix && !configuration.cardholderName) {
            NSCAssert(NO, @"Your PKAddPaymentPassRequestConfiguration must provide either a cardholderName or a primaryAccountSuffix.");
        }
    }
    return self;
}

- (instancetype)initWithNibName:(__unused NSString *)nibNameOrNil bundle:(__unused NSBundle *)nibBundleOrNil {
    return [self initWithRequestConfiguration:[PKAddPaymentPassRequestConfiguration new] delegate:nil];
}

- (instancetype)initWithCoder:(__unused NSCoder *)aDecoder {
    return [self initWithRequestConfiguration:[PKAddPaymentPassRequestConfiguration new] delegate:nil];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    
    UINavigationBar *navBar = [[UINavigationBar alloc] init];
    [self.view addSubview:navBar];
    navBar.translucent = NO;
    navBar.backgroundColor = [UIColor whiteColor];
    navBar.items = @[self.navigationItem];
    navBar.translatesAutoresizingMaskIntoConstraints = NO;
    [navBar.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active  = YES;
    [navBar.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    if (@available(iOS 11.0, *)) {
        [navBar.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor].active = YES;
    } else {
        [navBar.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = YES;
    }

    UILabel *contentLabel = [[UILabel alloc] initWithFrame:CGRectZero];
    self.contentLabel = contentLabel;
    contentLabel.textAlignment = NSTextAlignmentCenter;
    contentLabel.textColor = [UIColor blackColor];
    contentLabel.numberOfLines = 0;
    contentLabel.font = [UIFont systemFontOfSize:18];
    contentLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:contentLabel];
    [contentLabel.topAnchor constraintEqualToAnchor:navBar.bottomAnchor].active = YES;
    [contentLabel.leftAnchor constraintEqualToAnchor:self.view.leftAnchor constant:10.0f].active  = YES;
    [contentLabel.rightAnchor constraintEqualToAnchor:self.view.rightAnchor constant:-10.0f].active = YES;
    [contentLabel.heightAnchor constraintEqualToConstant:150].active = YES;
    
    NSMutableArray *pairs = [NSMutableArray array];
    if (self.configuration.cardholderName) {
        [pairs addObject:@[@"Name", self.configuration.cardholderName]];
    }
    if (self.configuration.primaryAccountSuffix) {
        [pairs addObject:@[@"Card Number", [NSString stringWithFormat:@"···· %@", self.configuration.primaryAccountSuffix]]];
    }
    NSMutableArray *rows = [NSMutableArray array];
    for (NSArray *pair in pairs) {
        UILabel *left = [[UILabel alloc] init];
        left.text = pair[0];
        left.textAlignment = NSTextAlignmentLeft;
        [left setFont:[UIFont boldSystemFontOfSize:16]];
        UILabel *right = [[UILabel alloc] init];
        right.text = pair[1];
        right.textAlignment = NSTextAlignmentLeft;
        right.textColor = [UIColor lightGrayColor];
        UIStackView *row = [[UIStackView alloc] initWithArrangedSubviews:@[left, right]];
        row.axis = UILayoutConstraintAxisHorizontal;
        row.distribution = UIStackViewDistributionFillEqually;
        row.alignment = UIStackViewAlignmentFill;
        row.translatesAutoresizingMaskIntoConstraints = NO;
        [rows addObject:row];
    }
    UIStackView *pairsTable = [[UIStackView alloc] initWithArrangedSubviews:rows];
    pairsTable.layoutMarginsRelativeArrangement = YES;
    pairsTable.layoutMargins = UIEdgeInsetsMake(20, 20, 20, 20);
    pairsTable.axis = UILayoutConstraintAxisVertical;
    pairsTable.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:pairsTable];

    [pairsTable.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active  = YES;
    [pairsTable.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    [pairsTable.topAnchor constraintEqualToAnchor:contentLabel.bottomAnchor].active = YES;
    [pairsTable.heightAnchor constraintEqualToConstant:(rows.count * 50)].active = YES;
    [self setState:STPFakeAddPaymentPassViewControllerStateInitial];
}

- (void)setState:(STPFakeAddPaymentPassViewControllerState)state {
    _state = state;
    UIBarButtonItem *cancelItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(cancel:)];
    UIButton *nextButton = [UIButton buttonWithType:UIButtonTypeSystem];
    [nextButton addTarget:self action:@selector(next:) forControlEvents:UIControlEventTouchUpInside];
    UIActivityIndicatorView *indicatorView = nil;
    #ifdef __IPHONE_13_0
            if (@available(iOS 13.0, *)) {
                indicatorView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleMedium];
            } else {
    #endif
    #if !(defined(TARGET_OS_MACCATALYST) && (TARGET_OS_MACCATALYST != 0))
                indicatorView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleGray];
    #endif
    #ifdef __IPHONE_13_0
            }
    #endif
    [indicatorView startAnimating];
    UIBarButtonItem *loadingItem = [[UIBarButtonItem alloc] initWithCustomView:indicatorView];
    [nextButton setTitle:STPNonLocalizedString(@"Next") forState:UIControlStateNormal];
    UIBarButtonItem *nextItem = [[UIBarButtonItem alloc] initWithCustomView:nextButton];
    UIBarButtonItem *doneItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(done:)];
    
    switch (state) {
        case STPFakeAddPaymentPassViewControllerStateInitial:
            self.contentLabel.text = STPNonLocalizedString(@"This class simulates the delegate methods that PKAddPaymentPassViewController will call in your app. Press next to continue.");
            self.navigationItem.leftBarButtonItem = cancelItem;
            self.navigationItem.rightBarButtonItem = nextItem;
            break;
        case STPFakeAddPaymentPassViewControllerStateLoading:
            self.contentLabel.text = STPNonLocalizedString(@"Fetching encrypted card details...");
            cancelItem.enabled = NO;
            self.navigationItem.leftBarButtonItem = cancelItem;
            self.navigationItem.rightBarButtonItem = loadingItem;
            break;
        case STPFakeAddPaymentPassViewControllerStateError:
            self.contentLabel.text = STPNonLocalizedString([@"Error: " stringByAppendingString:self.errorText]);
            doneItem.enabled = NO;
            self.navigationItem.leftBarButtonItem = cancelItem;
            self.navigationItem.rightBarButtonItem = doneItem;
            break;
        case STPFakeAddPaymentPassViewControllerStateSuccess:
            self.contentLabel.text = STPNonLocalizedString(@"Success! In production, your card would now have been added to your Apple Pay wallet. Your app's success callback will be triggered when the user presses 'Done'.");
            cancelItem.enabled = NO;
            self.navigationItem.leftBarButtonItem = cancelItem;
            self.navigationItem.rightBarButtonItem = doneItem;
    }
}

- (void)cancel:(__unused id)sender {
    [self.delegate addPaymentPassViewController:(PKAddPaymentPassViewController *)self didFinishAddingPaymentPass:nil error:[NSError errorWithDomain:PKPassKitErrorDomain code:PKAddPaymentPassErrorUserCancelled userInfo:nil]];
}

- (void)next:(__unused id)sender {
    [self setState:STPFakeAddPaymentPassViewControllerStateLoading];
    NSArray *certificates = @[
                     [@"cert1" dataUsingEncoding:NSUTF8StringEncoding],
                     [@"cert2" dataUsingEncoding:NSUTF8StringEncoding],
                     ];
    NSData *nonce = [@"nonce" dataUsingEncoding:NSUTF8StringEncoding];
    NSData *nonceSignature = [@"nonceSignature" dataUsingEncoding:NSUTF8StringEncoding];

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(10 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        if (self.state == STPFakeAddPaymentPassViewControllerStateLoading) {
            self.errorText = @"You exceeded the timeout of 10 seconds to call the request completion handler. Please check your PKAddPaymentPassViewControllerDelegate implementation, and make sure you are calling the `completionHandler` in `addPaymentPassViewController:generateRequestWithCertificateChain:nonce:nonceSignature:completionHandler`.";
            [self setState:STPFakeAddPaymentPassViewControllerStateError];
        }
    });
    [self.delegate addPaymentPassViewController:(PKAddPaymentPassViewController *)self
            generateRequestWithCertificateChain:certificates
                                          nonce:nonce
                                 nonceSignature:nonceSignature
                              completionHandler:^(PKAddPaymentPassRequest * _Nonnull request) {
                                  if (self.state == STPFakeAddPaymentPassViewControllerStateLoading) {
                                      NSString *contents;
                                      if (request.encryptedPassData) {
                                          contents = [[NSString alloc] initWithData:request.encryptedPassData encoding:NSUTF8StringEncoding];
                                      }
                                      if (request.stp_error) {
                                          NSString *error = request.stp_error.userInfo[STPErrorMessageKey];
                                          if (!error) {
                                              error = request.stp_error.userInfo[NSLocalizedDescriptionKey];
                                          }
                                          self.errorText = error;
                                          [self setState:STPFakeAddPaymentPassViewControllerStateError];
                                      }
                                      // This specific string is returned by the Stripe API in testmode.
                                      else if ([contents isEqualToString:@"TESTMODE_CONTENTS"]){
                                              [self setState:STPFakeAddPaymentPassViewControllerStateSuccess];
                                      } else {
                                          self.errorText = @"Your server response contained the wrong encrypted card details. Please ensure that you are not modifying the response from the Stripe API in any way, and that your request is in testmode.";
                                          [self setState:STPFakeAddPaymentPassViewControllerStateError];
                                      }
                                  }
}];
}

- (void)done:(__unused id)sender {
    PKPaymentPass *pass = [PKPaymentPass new];
    [self.delegate addPaymentPassViewController:(PKAddPaymentPassViewController *)self didFinishAddingPaymentPass:pass error:nil];
}

@end

File: Stripe/StripeiOS/Source/STPFakeAddPaymentPassViewController.swift
//
//  STPFakeAddPaymentPassViewController.swift
//  StripeiOS
//
//  Created by Jack Flintermann on 9/28/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

import PassKit
@_spi(STP) import StripeCore
import UIKit

/// This class is a piece of fake UI that is intended to mimic `PKAddPaymentPassViewController`. That class is restricted to apps with a special entitlement from Apple, and as such can be difficult to build and test against. This class implements the same public API as `PKAddPaymentPassViewController`, and can be used to develop against the Stripe API in *testmode only*. (Obviously it will not actually place cards into the user's Apple Pay wallet either.) When it's time to go to production, you may simply replace all references to `STPFakeAddPaymentPassViewController` in your app with `PKAddPaymentPassViewController` and it will continue to function. For more information on developing against this API, please see https://stripe.com/docs/issuing/cards/digital-wallets .
public class STPFakeAddPaymentPassViewController: UIViewController {
    /// @see PKAddPaymentPassViewController
    @objc
    public class func canAddPaymentPass() -> Bool {
        return true
    }

    /// @see PKAddPaymentPassViewController
    @objc(initWithRequestConfiguration:delegate:)
    public required init?(
        requestConfiguration configuration: PKAddPaymentPassRequestConfiguration,
        delegate: PKAddPaymentPassViewControllerDelegate?
    ) {
        super.init(nibName: nil, bundle: nil)
        assert(delegate != nil, "Invalid parameter not satisfying: delegate != nil")
        state = .initial
        self.delegate = delegate
        self.configuration = configuration
        if configuration.primaryAccountSuffix == nil && configuration.cardholderName == nil {
            assert(
                false,
                "Your PKAddPaymentPassRequestConfiguration must provide either a cardholderName or a primaryAccountSuffix."
            )
        }
    }

    /// @see PKAddPaymentPassViewController
    @objc public weak var delegate: PKAddPaymentPassViewControllerDelegate?
    private var configuration: PKAddPaymentPassRequestConfiguration?

    private var _state: STPFakeAddPaymentPassViewControllerState = .initial
    private var state: STPFakeAddPaymentPassViewControllerState {
        get {
            _state
        }
        set(state) {
            _state = state
            let cancelItem = UIBarButtonItem(
                barButtonSystemItem: .cancel,
                target: self,
                action: #selector(cancel(_:))
            )
            let nextButton = UIButton(type: .system)
            nextButton.addTarget(self, action: #selector(next(_:)), for: .touchUpInside)
            let indicatorView = UIActivityIndicatorView(style: .medium)
            indicatorView.startAnimating()
            let loadingItem = UIBarButtonItem(customView: indicatorView)
            nextButton.setTitle(STPNonLocalizedString("Next"), for: .normal)
            let nextItem = UIBarButtonItem(customView: nextButton)
            let doneItem = UIBarButtonItem(
                barButtonSystemItem: .done,
                target: self,
                action: #selector(done(_:))
            )

            switch state {
            case .initial:
                contentLabel?.text = STPNonLocalizedString(
                    "This class simulates the delegate methods that PKAddPaymentPassViewController will call in your app. Press next to continue."
                )
                navigationItem.leftBarButtonItem = cancelItem
                navigationItem.rightBarButtonItem = nextItem
            case .loading:
                contentLabel?.text = STPNonLocalizedString("Fetching encrypted card details...")
                cancelItem.isEnabled = false
                navigationItem.leftBarButtonItem = cancelItem
                navigationItem.rightBarButtonItem = loadingItem
            case .error:
                contentLabel?.text = STPNonLocalizedString("Error: " + (errorText ?? ""))
                doneItem.isEnabled = false
                navigationItem.leftBarButtonItem = cancelItem
                navigationItem.rightBarButtonItem = doneItem
            case .success:
                contentLabel?.text = STPNonLocalizedString(
                    "Success! In production, your card would now have been added to your Apple Pay wallet. Your app's success callback will be triggered when the user presses 'Done'."
                )
                cancelItem.isEnabled = false
                navigationItem.leftBarButtonItem = cancelItem
                navigationItem.rightBarButtonItem = doneItem
            }
        }
    }
    private var contentLabel: UILabel?
    private var errorText: String?

    /// :nodoc:
    @objc public convenience override init(
        nibName nibNameOrNil: String?,
        bundle nibBundleOrNil: Bundle?
    ) {
        self.init(requestConfiguration: PKAddPaymentPassRequestConfiguration(), delegate: nil)!
    }

    /// :nodoc:
    @objc public required convenience init?(
        coder aDecoder: NSCoder
    ) {
        self.init(requestConfiguration: PKAddPaymentPassRequestConfiguration(), delegate: nil)
    }

    /// :nodoc:
    @objc
    public override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor.white

        let navBar = UINavigationBar()
        view.addSubview(navBar)
        navBar.isTranslucent = false
        navBar.backgroundColor = UIColor.white
        navBar.items = [navigationItem]
        navBar.translatesAutoresizingMaskIntoConstraints = false
        navBar.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        navBar.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true
        navBar.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor).isActive = true

        let contentLabel = UILabel(frame: CGRect.zero)
        self.contentLabel = contentLabel
        contentLabel.textAlignment = .center
        contentLabel.textColor = UIColor.black
        contentLabel.numberOfLines = 0
        contentLabel.font = UIFont.systemFont(ofSize: 18)
        contentLabel.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(contentLabel)
        contentLabel.topAnchor.constraint(equalTo: navBar.bottomAnchor).isActive = true
        contentLabel.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 10.0).isActive = true
        contentLabel.rightAnchor.constraint(equalTo: view.rightAnchor, constant: -10.0).isActive =
            true
        contentLabel.heightAnchor.constraint(equalToConstant: 150).isActive = true

        var pairs: [AnyHashable] = []
        if configuration?.cardholderName != nil {
            pairs.append(["Name", configuration?.cardholderName])
        }
        if configuration?.primaryAccountSuffix != nil {
            pairs.append([
                "Card Number",
                "···· \(configuration?.primaryAccountSuffix ?? "")",
            ])
        }
        var rows: [AnyHashable] = []
        for pair in pairs {
            guard let pair = pair as? [AnyHashable] else {
                continue
            }
            let left = UILabel()
            left.text = pair[0] as? String
            left.textAlignment = .left
            left.font = UIFont.boldSystemFont(ofSize: 16)
            let right = UILabel()
            right.text = pair[1] as? String
            right.textAlignment = .left
            right.textColor = UIColor.lightGray
            let row = UIStackView(arrangedSubviews: [left, right])
            row.axis = .horizontal
            row.distribution = .fillEqually
            row.alignment = .fill
            row.translatesAutoresizingMaskIntoConstraints = false
            rows.append(row)
        }
        var pairsTable: UIStackView?
        if let rows = rows as? [UIView] {
            pairsTable = UIStackView(arrangedSubviews: rows)
        }
        pairsTable?.isLayoutMarginsRelativeArrangement = true
        pairsTable?.layoutMargins = UIEdgeInsets(top: 20, left: 20, bottom: 20, right: 20)
        pairsTable?.axis = .vertical
        pairsTable?.translatesAutoresizingMaskIntoConstraints = false
        if let pairsTable = pairsTable {
            view.addSubview(pairsTable)
        }

        pairsTable?.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        pairsTable?.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true
        pairsTable?.topAnchor.constraint(equalTo: contentLabel.bottomAnchor).isActive = true
        pairsTable?.heightAnchor.constraint(equalToConstant: CGFloat((rows.count * 50))).isActive =
            true
        state = .initial
    }

    @objc func cancel(_ sender: Any?) {
        let castedVC = unsafeBitCast(self, to: PKAddPaymentPassViewController.self)
        delegate?.addPaymentPassViewController(
            castedVC,
            didFinishAdding: nil,
            error: NSError(
                domain: PKPassKitErrorDomain,
                code: PKAddPaymentPassError.userCancelled.rawValue,
                userInfo: nil
            )
        )
    }

    @objc func next(_ sender: Any?) {
        state = .loading
        let certificates = [
            "cert1".data(using: .utf8),
            "cert2".data(using: .utf8),
        ]
        let nonce = "nonce".data(using: .utf8)
        let nonceSignature = "nonceSignature".data(using: .utf8)

        DispatchQueue.main.asyncAfter(
            deadline: DispatchTime.now() + Double(Int64(10 * Double(NSEC_PER_SEC)))
                / Double(NSEC_PER_SEC),
            execute: {
                if self.state == .loading {
                    self.errorText =
                        "You exceeded the timeout of 10 seconds to call the request completion handler. Please check your PKAddPaymentPassViewControllerDelegate implementation, and make sure you are calling the `completionHandler` in `addPaymentPassViewController:generateRequestWithCertificateChain:nonce:nonceSignature:completionHandler`."
                    self.state = .error
                }
            }
        )
        if let nonce = nonce, let nonceSignature = nonceSignature {
            let castedVC = unsafeBitCast(self, to: PKAddPaymentPassViewController.self)
            delegate?.addPaymentPassViewController(
                castedVC,
                generateRequestWithCertificateChain: certificates.compactMap { $0 },
                nonce: nonce,
                nonceSignature: nonceSignature,
                completionHandler: { request in
                    if self.state == .loading {
                        var contents: String?
                        if request.encryptedPassData != nil {
                            if let encryptedPassData1 = request.encryptedPassData {
                                contents = String(data: encryptedPassData1, encoding: .utf8)
                            }
                        }
                        if request.stp_error != nil {
                            var error =
                                (request.stp_error as NSError?)?.userInfo[STPError.errorMessageKey]
                                as? String
                            if error == nil {
                                error =
                                    (request.stp_error as NSError?)?.userInfo[
                                        NSLocalizedDescriptionKey
                                    ] as? String
                            }
                            self.errorText = error
                            self.state = .error
                        } else if contents == "TESTMODE_CONTENTS" {
                            self.state = .success
                        } else {
                            self.errorText =
                                "Your server response contained the wrong encrypted card details. Please ensure that you are not modifying the response from the Stripe API in any way, and that your request is in testmode."
                            self.state = .error
                        }
                    }
                }
            )
        }
    }

    @objc func done(_ sender: Any?) {
        let pass = PKPaymentPass()
        let castedVC = unsafeBitCast(self, to: PKAddPaymentPassViewController.self)
        delegate?.addPaymentPassViewController(castedVC, didFinishAdding: pass, error: nil)
    }
}

enum STPFakeAddPaymentPassViewControllerState: Int {
    case initial
    case loading
    case error
    case success
}
